application/application.go

```diff
@@ -0,0 +1,101 @@
+package application
```

Comment:
@josephroberts an overall arch point. I think you're chosing to use interfaces really early and then trying to fight against the type system to make things work.

If you have a core data structure that you are using for each device (and don't need to have additional state for devices), you can then deal with this struct as necessary without having to any dancing around with getters or marshalling/unmarshalling tricks, and then convert to the specific type when required.

If you have `type BlahDevice Device`, you can define unique methods on `BlahDevice` and keep everything that already exists from the `Device` type, and simply cast `Device` to `BlahDevice` when you need it.

I've created a [playground example](https://play.golang.org/p/8h9O9sAFNn) to demonstrate what I mean.

If you _definitely_ need to have additional state for some devices at some point then this isn't going to work so well and the current getter functionality + an embedded value works better, but if not then this will make life easier.